File: bindings/python/python/matter_sdk/facade.py
"""Curated typed façades over :meth:`MatterClient.tx`.

Each method only shapes typed arguments into ``call_params``, keyed by the
runtime's own argument names, and delegates: no encoding or chain access of its
own. Anything not here is one ``client.tx(...)`` away.

Names resolve at call time from live metadata, so a runtime rename fails at the call.

The surface is pinned by ``testvectors/facade_calls.json``; ``tests/test_facade.py``
checks it both ways.
"""

from typing import Any, Dict, List, Optional


def _account_arg(account: Any) -> Any:
    """Render raw account bytes (as ``lookup`` returns) as the ``0x`` hex
    scalecodec accepts; pass anything else through."""
    if isinstance(account, (bytes, bytearray)):
        return "0x" + bytes(account).hex()
    return account

__all__ = [
    "SecretsFacade",
    "DeploymentsFacade",
    "ResourcesFacade",
    "StakingFacade",
    "OrgsFacade",
    "KeysFacade",
]


class _Facade:
    """Common base: holds the client and forwards a named call."""

    __slots__ = ("_client",)

    def __init__(self, client) -> None:
        self._client = client

    def _tx(self, pallet: str, call: str, params: Dict[str, Any]):
        return self._client.tx(pallet, call, params)


class SecretsFacade(_Facade):
    """Secrets: store, rotate, share, and delete secrets."""

    def store(self, payload, epoch: int, label, aad):
        """Publish a sealed envelope.

        ``payload`` is the tuple ``encrypt`` returns. The chain-assigned id is in
        the receipt's ``Secrets.SecretStored`` event.
        """
        from .calls import store_secret

        return self._tx("Secrets", "store_secret", store_secret(payload, epoch, label, aad))

    def rotate(self, secret_id: int, payload, epoch: int, aad):
        """Re-seal an existing secret in place under the current epoch."""
        from .calls import rotate_secret

        return self._tx(
            "Secrets", "rotate_secret", rotate_secret(secret_id, payload, epoch, aad)
        )

    def grant(self, secret_id: int, target: dict):
        """Authorize a principal to request decryption.

        ``target`` comes from :func:`~matter_sdk.calls.grant_to_user` or
        :func:`~matter_sdk.calls.grant_to_deployment`.
        """
        from .calls import grant_access

        return self._tx("Secrets", "grant_access", grant_access(secret_id, target))

    def revoke(self, secret_id: int, target: dict):
        """Withdraw a grant.

        The target must match the grant exactly, or this is a no-op on chain.
        """
        from .calls import revoke_access

        return self._tx("Secrets", "revoke_access", revoke_access(secret_id, target))

    def delete(self, secret_id: int):
        """Delete a secret and every grant on it. Owner only, and irreversible."""
        from .calls import delete_secret

        return self._tx("Secrets", "delete_secret", delete_secret(secret_id))


class DeploymentsFacade(_Facade):
    """Deployments (``pallet-jobs``): request compute, wire networking, bind secrets."""

    def request(self, request: Any):
        """Request a deployment.

        ``request`` is the runtime's ``ResourceRequest``, passed through as-is.
        """
        return self._tx("Jobs", "request_deployment", {"request": request})

    def cancel(self, deployment: int):
        """Cancel a deployment."""
        return self._tx("Jobs", "cancel_deployment", {"deployment": int(deployment)})

    def set_secret_ref(self, deployment: int, secret_ref: Optional[int]):
        """Point a deployment at a secret, or clear it with ``None``.

        The assigned resource is authorized to decrypt whatever ``secret_ref`` names.
        """
        return self._tx(
            "Jobs",
            "set_deployment_secret_ref",
            {
                "deployment": int(deployment),
                "secret_ref": None if secret_ref is None else int(secret_ref),
            },
        )

    def set_env(self, deployment: int, env_vars: Optional[Any]):
        """Set or clear a deployment's **plaintext** environment variables.

        Anything sensitive belongs in a sealed secret referenced by
        :meth:`set_secret_ref`, not here.
        """
        return self._tx(
            "Jobs",
            "set_deployment_env",
            {"deployment": int(deployment), "env_vars": env_vars},
        )

    def register_wg_peer(self, deployment: int, user_wg_pubkey: bytes, pq_ciphertext: bytes):
        """Register a WireGuard peer public key against a deployment.

        ``pq_ciphertext`` is the ML-KEM-768 ciphertext (1088 bytes) the peer
        encapsulated to the provider's on-chain ML-KEM key
        (``OverlayNetworks.PqKemPubkeys``); the provider decapsulates it to
        derive the tunnel's post-quantum preshared key. Requires runtime spec >= 330.
        """
        from .hexutil import to_hex

        return self._tx(
            "Jobs",
            "register_wg_peer",
            {
                "deployment": int(deployment),
                "user_wg_pubkey": to_hex(user_wg_pubkey),
                "pq_ciphertext": to_hex(pq_ciphertext),
            },
        )


class ResourcesFacade(_Facade):
    """Resources: register capacity, price it, control who may use it."""

    def register(self, resource_id: str, ownership_proof: Any, name: str):
        """Register a resource you operate."""
        return self._tx(
            "Resources",
            "register_resource",
            {"resource_id": resource_id, "ownership_proof": ownership_proof, "name": name},
        )

    def update_sku(self, uuid: int, sku: Any):
        """Publish or update a SKU's pricing."""
        return self._tx("Resources", "update_sku", {"uuid": int(uuid), "sku": sku})

    def report_capacity(self, capacity: Any):
        """Report current capacity."""
        return self._tx("Resources", "report_capacity", {"capacity": capacity})

    def set_privacy(self, resource_id: str, is_private: bool):
        """Make a resource private (whitelist-only) or public."""
        return self._tx(
            "Resources",
            "set_resource_privacy",
            {"resource_id": resource_id, "is_private": bool(is_private)},
        )

    def allow(self, resource_id: str, user: str):
        """Allow an account to use a private resource."""
        return self._tx(
            "Resources", "add_to_whitelist", {"resource_id": resource_id, "user": user}
        )

    def disallow(self, resource_id: str, user: str):
        """Withdraw a private resource's whitelist entry."""
        return self._tx(
            "Resources", "remove_from_whitelist", {"resource_id": resource_id, "user": user}
        )


class StakingFacade(_Facade):
    """Staking on MatterChain: the standard FRAME staking surface.

    Amounts are **plancks**; use :meth:`MatterClient.parse_amount` rather than a
    hand-written exponent. ``pallet-staking-gateway`` (Ethereum meta-transactions)
    is intentionally absent.
    """

    def bond(self, value: int, payee: Any):
        """Bond funds and set a reward destination, e.g. ``{"Staked": None}``."""
        return self._tx("Staking", "bond", {"value": int(value), "payee": payee})

    def bond_extra(self, max_additional: int):
        """Add to an existing bond."""
        return self._tx(
            "Staking", "bond_extra", {"max_additional": int(max_additional)}
        )

    def unbond(self, value: int):
        """Schedule an unbond.

        Funds stay locked until the unbonding period elapses and
        :meth:`withdraw_unbonded` is called.
        """
        return self._tx("Staking", "unbond", {"value": int(value)})

    def withdraw_unbonded(self, num_slashing_spans: int):
        """Move unlocked funds back to free balance."""
        return self._tx(
            "Staking", "withdraw_unbonded", {"num_slashing_spans": int(num_slashing_spans)}
        )

    def nominate(self, targets: List[str]):
        """Nominate validators, by SS58 address."""
        return self._tx("Staking", "nominate", {"targets": list(targets)})

    def chill(self):
        """Stop nominating or validating."""
        return self._tx("Staking", "chill", {})

    def join_pool(self, amount: int, pool_id: int):
        """Join a nomination pool with ``amount`` plancks."""
        return self._tx(
            "NominationPools", "join", {"amount": int(amount), "pool_id": int(pool_id)}
        )

    def claim_pool_payout(self):
        """Claim accrued nomination-pool rewards."""
        return self._tx("NominationPools", "claim_payout", {})


class OrgsFacade(_Facade):
    """Organizations and budgets: membership, projects, and who may spend or decrypt."""

    def create(self):
        """Create an organization.

        The runtime derives the org id from the signer and a sequence number.
        """
        return self._tx("Organizations", "create_org", {})

    def add_member(self, org: Any, who: str, role: Any):
        """Add a member with a role."""
        return self._tx(
            "Organizations", "add_member", {"org": org, "who": who, "role": role}
        )

    def remove_member(self, org: Any, who: str):
        """Remove a member."""
        return self._tx("Organizations", "remove_member", {"org": org, "who": who})

    def allot(self, org: Any, project: Any, amount: int):
        """Allot budget from an org treasury to a project, in plancks."""
        return self._tx(
            "Budgets", "allot", {"org": org, "project": project, "amount": int(amount)}
        )

    def authorize_secrets_agent(self, org: Any, project: Any, who: str):
        """Authorize an account to decrypt a project's secrets.

        The org-scoped analogue of ``secrets.grant_access``.
        """
        return self._tx(
            "Budgets",
            "authorize_project_secrets_agent",
            {"org": org, "project": project, "who": who},
        )

    def revoke_secrets_agent(self, org: Any, project: Any, who: str):
        """Withdraw a project secrets-agent authorization."""
        return self._tx(
            "Budgets",
            "revoke_project_secrets_agent",
            {"org": org, "project": project, "who": who},
        )


class KeysFacade(_Facade):
    """Minting and revoking member-tied API keys (``pallet-budgets``' roster calls).

    **Member-signed.** The runtime never admits these to a key, so a key cannot
    widen its own authority; use a client built on a seed or keypair. ``authorize``
    from a delegated client raises before submitting.
    """

    def authorize(self, key: Any, scopes: Any):
        """Register ``key`` as an API key acting for the signer, with ``scopes``.

        An upsert, so re-scoping a live key is this same call. ``scopes`` is a
        :class:`~matter_sdk.scopes.ScopeSet` or its raw ``u32`` bits.
        """
        bits = scopes.bits if hasattr(scopes, "bits") else int(scopes)
        return self._tx(
            "Budgets", "authorize_agent_key", {"key": _account_arg(key), "scopes": bits}
        )

    def revoke(self, key: Any):
        """Revoke ``key``, cutting off its authority and committee decrypt rights
        from the next request onward."""
        return self._tx("Budgets", "revoke_agent_key", {"key": _account_arg(key)})

    def lookup(self, key: bytes):
        """Who ``key`` acts for and what it may do, or ``None`` if unregistered.

        A read, so it needs no signer and works on a read-only client.
        """
        from .scopes import ScopeSet

        resolved = self._client.chain.agent_key(key)
        if resolved is None:
            return None
        principal, bits = resolved
        return (principal, ScopeSet.from_bits(bits))

File: packages/typescript/src/facade.ts
// Typed convenience wrappers over MatterClient.tx: they shape arguments and
// delegate, with no encoding or chain access of their own. Anything not here is
// one `client.tx(...)` away.

import type { Aad, EncryptedSecret, GrantTarget, StoreSecret } from "@openmatter-network/matter-sdk-core";
import { aadBytes } from "@openmatter-network/matter-sdk-core";

import type { TxReceipt } from "./backend.js";
import type { ScopeSet } from "./scopes.js";

/** What a façade needs from the client. */
export interface FacadeHost {
  tx(pallet: string, call: string, args: unknown[]): Promise<TxReceipt>;
  agentKey?(key: Uint8Array): Promise<readonly [Uint8Array, ScopeSet] | undefined>;
}

/** The four-blob sealed envelope as the runtime's `EncryptedSecret` composite. */
function envelope(payload: EncryptedSecret): unknown {
  return {
    binding_id: payload.bindingId,
    capsule: payload.capsule,
    proof: payload.proof,
    ct: payload.ct,
  };
}

/** Secrets: store, rotate, share, and delete secrets. */
export class SecretsFacade {
  constructor(private readonly host: FacadeHost) {}

  /**
   * Publish a sealed envelope. The chain-assigned id is in the
   * `Secrets.SecretStored` event on the returned receipt.
   */
  async store(args: StoreSecret): Promise<TxReceipt> {
    return this.host.tx("Secrets", "store_secret", [
      envelope(args.payload),
      args.epoch,
      args.label,
      args.aad,
    ]);
  }

  /** Re-seal an existing secret in place under the current epoch. */
  async rotate(
    secretId: bigint,
    payload: EncryptedSecret,
    epoch: number,
    aad: Aad,
  ): Promise<TxReceipt> {
    return this.host.tx("Secrets", "rotate_secret", [
      secretId,
      envelope(payload),
      epoch,
      aadBytes(aad),
    ]);
  }

  /** Authorize a principal to request decryption. */
  async grant(secretId: bigint, target: GrantTarget): Promise<TxReceipt> {
    return this.host.tx("Secrets", "grant_access", [secretId, target]);
  }

  /** Withdraw a grant. `target` must match the grant exactly, else a no-op on chain. */
  async revoke(secretId: bigint, target: GrantTarget): Promise<TxReceipt> {
    return this.host.tx("Secrets", "revoke_access", [secretId, target]);
  }

  /** Delete a secret and every grant on it. Owner only, and irreversible. */
  async delete(secretId: bigint): Promise<TxReceipt> {
    return this.host.tx("Secrets", "delete_secret", [secretId]);
  }
}

/** Deployments (`pallet-jobs`): request compute, wire networking, bind secrets. */
export class DeploymentsFacade {
  constructor(private readonly host: FacadeHost) {}

  /** Request a deployment. `request` is the runtime's `ResourceRequest`, passed through. */
  async request(request: unknown): Promise<TxReceipt> {
    return this.host.tx("Jobs", "request_deployment", [request]);
  }

  async cancel(deployment: bigint): Promise<TxReceipt> {
    return this.host.tx("Jobs", "cancel_deployment", [deployment]);
  }

  /**
   * Point a deployment at a secret, or clear it with `null`. The assigned
   * resource may then decrypt whatever `secretRef` names.
   */
  async setSecretRef(deployment: bigint, secretRef: bigint | null): Promise<TxReceipt> {
    return this.host.tx("Jobs", "set_deployment_secret_ref", [deployment, secretRef]);
  }

  /**
   * Set or clear a deployment's **plaintext** environment variables. Anything
   * sensitive belongs in a sealed secret referenced by {@link setSecretRef}.
   */
  async setEnv(deployment: bigint, env: unknown | null): Promise<TxReceipt> {
    return this.host.tx("Jobs", "set_deployment_env", [deployment, env]);
  }

  /**
   * Register a WireGuard peer public key against a deployment. Requires runtime
   * spec >= 330.
   *
   * `pqCiphertext`: the 1088-byte ML-KEM-768 ciphertext encapsulated to the
   * provider's `OverlayNetworks.PqKemPubkeys` key; it derives the tunnel's PSK.
   */
  async registerWgPeer(
    deployment: bigint,
    pubkey: Uint8Array,
    pqCiphertext: Uint8Array,
  ): Promise<TxReceipt> {
    return this.host.tx("Jobs", "register_wg_peer", [deployment, pubkey, pqCiphertext]);
  }
}

/** Resources: register capacity, price it, control who may use it. */
export class ResourcesFacade {
  constructor(private readonly host: FacadeHost) {}

  async register(resource: Uint8Array, ownershipProof: unknown, name: string): Promise<TxReceipt> {
    return this.host.tx("Resources", "register_resource", [resource, ownershipProof, name]);
  }

  /** Publish or update a SKU's pricing. */
  async updateSku(uuid: bigint, sku: unknown): Promise<TxReceipt> {
    return this.host.tx("Resources", "update_sku", [uuid, sku]);
  }

  async reportCapacity(capacity: unknown): Promise<TxReceipt> {
    return this.host.tx("Resources", "report_capacity", [capacity]);
  }

  /** Make a resource private (whitelist-only) or public. */
  async setPrivacy(resource: Uint8Array, isPrivate: boolean): Promise<TxReceipt> {
    return this.host.tx("Resources", "set_resource_privacy", [resource, isPrivate]);
  }

  /** Allow an account to use a private resource. */
  async allow(resource: Uint8Array, user: Uint8Array): Promise<TxReceipt> {
    return this.host.tx("Resources", "add_to_whitelist", [resource, user]);
  }

  /** Withdraw a private resource's whitelist entry. */
  async disallow(resource: Uint8Array, user: Uint8Array): Promise<TxReceipt> {
    return this.host.tx("Resources", "remove_from_whitelist", [resource, user]);
  }
}

/**
 * Staking on MatterChain, including nomination pools. Amounts are **plancks**;
 * use `client.parseAmount()` rather than a hand-written exponent.
 */
export class StakingFacade {
  constructor(private readonly host: FacadeHost) {}

  /** Bond funds and set a reward destination, e.g. `{ Staked: null }`. */
  async bond(value: bigint, payee: unknown): Promise<TxReceipt> {
    return this.host.tx("Staking", "bond", [value, payee]);
  }

  async bondExtra(maxAdditional: bigint): Promise<TxReceipt> {
    return this.host.tx("Staking", "bond_extra", [maxAdditional]);
  }

  /**
   * Schedule an unbond. Funds stay locked until the unbonding period elapses and
   * {@link withdrawUnbonded} is called.
   */
  async unbond(value: bigint): Promise<TxReceipt> {
    return this.host.tx("Staking", "unbond", [value]);
  }

  /** Move unlocked funds back to free balance. */
  async withdrawUnbonded(numSlashingSpans: number): Promise<TxReceipt> {
    return this.host.tx("Staking", "withdraw_unbonded", [numSlashingSpans]);
  }

  /** Nominate validators, by 32-byte account id. */
  async nominate(targets: Uint8Array[]): Promise<TxReceipt> {
    return this.host.tx("Staking", "nominate", [targets.map((id) => ({ Id: id }))]);
  }

  async chill(): Promise<TxReceipt> {
    return this.host.tx("Staking", "chill", []);
  }

  /** Join a nomination pool with `amount` plancks. */
  async joinPool(amount: bigint, poolId: number): Promise<TxReceipt> {
    return this.host.tx("NominationPools", "join", [amount, poolId]);
  }

  async claimPoolPayout(): Promise<TxReceipt> {
    return this.host.tx("NominationPools", "claim_payout", []);
  }
}

/** Organizations and budgets: membership, projects, and who may spend or decrypt. */
export class OrgsFacade {
  constructor(private readonly host: FacadeHost) {}

  /** Create an organization; the runtime derives its id from the signer. */
  async create(): Promise<TxReceipt> {
    return this.host.tx("Organizations", "create_org", []);
  }

  async addMember(org: Uint8Array, who: Uint8Array, role: unknown): Promise<TxReceipt> {
    return this.host.tx("Organizations", "add_member", [org, who, role]);
  }

  async removeMember(org: Uint8Array, who: Uint8Array): Promise<TxReceipt> {
    return this.host.tx("Organizations", "remove_member", [org, who]);
  }

  /** Allot budget from an org treasury to a project, in plancks. */
  async allot(org: Uint8Array, project: Uint8Array, amount: bigint): Promise<TxReceipt> {
    return this.host.tx("Budgets", "allot", [org, project, amount]);
  }

  /** Authorize an account to decrypt a project's secrets. */
  async authorizeSecretsAgent(
    org: Uint8Array,
    project: Uint8Array,
    who: Uint8Array,
  ): Promise<TxReceipt> {
    return this.host.tx("Budgets", "authorize_project_secrets_agent", [org, project, who]);
  }

  /** Withdraw a project secrets-agent authorization. */
  async revokeSecretsAgent(
    org: Uint8Array,
    project: Uint8Array,
    who: Uint8Array,
  ): Promise<TxReceipt> {
    return this.host.tx("Budgets", "revoke_project_secrets_agent", [org, project, who]);
  }
}

/**
 * Minting and revoking member-tied API keys.
 *
 * **Member-signed only.** From a delegated client, `authorize` and `revoke`
 * throw `ClientError` `never-admitted` before submitting, so a key cannot widen
 * its own authority.
 */
export class KeysFacade {
  constructor(private readonly host: FacadeHost) {}

  /**
   * Register `key` as an API key acting for the signer, with `scopes`. Upsert:
   * re-scoping a live key is this same call.
   */
  async authorize(key: Uint8Array, scopes: ScopeSet): Promise<TxReceipt> {
    return this.host.tx("Budgets", "authorize_agent_key", [key, scopes.bits]);
  }

  /** Revoke `key`'s authority and committee decrypt rights from the next request on. */
  async revoke(key: Uint8Array): Promise<TxReceipt> {
    return this.host.tx("Budgets", "revoke_agent_key", [key]);
  }

  /**
   * Who `key` acts for and what it may do, or `undefined` if unregistered. Works
   * on a read-only client.
   */
  async lookup(key: Uint8Array): Promise<readonly [Uint8Array, ScopeSet] | undefined> {
    if (this.host.agentKey === undefined) {
      throw new Error("this backend cannot resolve an api key's grant");
    }
    return this.host.agentKey(key);
  }
}
